Add OpenBSD and NetBSD?

https://github.com/haskell/unix/pull/262

But we have no bindists so far.

I've been messing around with GHCup on OpenBSD (7.5), and these are the roadblocks to getting GHCup to compile and run:

  - `lzma-static` dependency's vendored `liblzma` doesn't compile on OpenBSD (though upstream `liblzma` does).
    - Solution 1: switch to `lzma`, which doesn't vendor `liblzma` (funnily enough, in `cabal.project.release`, we pass `+static` as a build flag constraint to `lzma` in some environments, but it doesn't seem to do anything with it, not even do a static build, which is what I expected; if you can explain why we pass it that build flag, I'd appreciate it).

      `libarchive` has a transitive dependency on `lzma-static`, so the above solution alone doesn't entirely solve the problem; you'd also have to enable the `tar` build flag to use `zip` and `tar` dependencies instead of `libarchive`.
    - Solution 2: generate the platform-specific `config.h` file for OpenBSD.

      `lzma-static` vendors `liblzma` from the `xz` source tree. Due to `liblzma` disallowing compilation on architectures other than 32- or 64-bit, and the inability to use the `sizeof()` macro during the preprocessing phase to test for this condition before runtime, it requires the use of a `configure` script to define `SIZEOF_SIZE_T`. `liblzma` upstream includes this `configure` script, but `lzma-static` opts instead to vendor platform-specific `config.h` header files, which it has a `gen-header.sh` script to generate (the script fetches an `xz` source tarball to use the `configure` script from it to and copies the generated `config.h` file).

      ```diff
      diff -ru a/gen-header.sh b/gen-header.sh
      --- a/gen-header.sh
      +++ b/gen-header.sh
      @@ -13,7 +13,7 @@ XZ_URL="https://tukaani.org/xz/${XZ_BALL}"
       
       [ -f "${XZ_BALL}" ] || curl -LO "${XZ_URL}"
       
      -[ -d "xz-${XZ_VER}" ] || tar xf "${XZ_BALL}"
      +[ -d "xz-${XZ_VER}" ] || bzcat "${XZ_BALL}" | tar xf -
       
       cd "xz-${XZ_VER}"
       ./configure
      @@ -69,6 +69,28 @@ case "$(uname -s)" in
                     cp config.h ../autoconf-freebsd-aarch64/config.h ;;
          esac
          ;;
      +	"OpenBSD"|"openbsd")
      +		case "$(uname -m)" in
      +			x86_64|amd64)
      +				if [ "$(getconf LONG_BIT)" = "32" ] ; then
      +					cp config.h ../autoconf-openbsd-i386/config.h
      +				else
      +					cp config.h ../autoconf-openbsd-x86_64/config.h
      +				fi
      +				;;
      +			i*86)
      +				cp config.h ../autoconf-openbsd-i386/config.h ;;
      +			armv7*)
      +				cp config.h ../autoconf-openbsd-arm/config.h ;;
      +			aarch64|arm64|armv8l)
      +				if [ "$(getconf LONG_BIT)" = "32" ] ; then
      +					cp config.h ../autoconf-openbsd-arm/config.h
      +				else
      +					cp config.h ../autoconf-openbsd-aarch64/config.h
      +				fi
      +				;;
      +		esac
      +		;;
        *) die "Unknown platform" ;;
       esac
       
      diff -ru a/lzma-static.cabal b/lzma-static.cabal
      --- a/lzma-static.cabal
      +++ b/lzma-static.cabal
      @@ -91,6 +91,15 @@ library
             include-dirs: autoconf-freebsd
           elif arch(aarch64)
             include-dirs: autoconf-freebsd-aarch64
      +  elif os(openbsd)
      +    if arch(x86_64)
      +      include-dirs: autoconf-openbsd-x86_64
      +    elif arch(i386)
      +      include-dirs: autoconf-openbsd-i386
      +    elif arch(arm)
      +      include-dirs: autoconf-openbsd-arm
      +    elif arch(aarch64)
      +      include-dirs: autoconf-openbsd-aarch64
         else
           build-depends: unbuildable<0
           buildable: False
      ```

      (I've only built on OpenBSD amd64.)
    - Solution 3: we can go further and use a generic fallback `config.h` file to make the vendored `liblzma` (which hasn't been updated since the initial commit) easily buildable on OpenBSD, and other arbitrary platforms which don't have their own `config.h` file. (We can still also use an OpenBSD-specific `config.h` file if we want.)

      ```diff
      diff -ru a/lzma-static.cabal b/lzma-static.cabal
      --- a/lzma-static.cabal
      +++ b/lzma-static.cabal
      @@ -92,8 +92,7 @@ library
           elif arch(aarch64)
             include-dirs: autoconf-freebsd-aarch64
         else
      -    build-depends: unbuildable<0
      -    buildable: False
      +    include-dirs: autoconf-generic
       
         c-sources:
           cbits/lzma_wrapper.c
      ```

      Most of the platform-specific definitions seem to be unnecessary (`liblzma` has fallback behaviour, and the generated `config.h` even comments out some definitions instead of defining them as `0`), but `SIZEOF_SIZE_T` is required and can't simply be set to a reasonable default.

      ```c
      // Incorrect(?) SIZE_MAX:
      //   - Interix headers typedef size_t to unsigned long,
      //     but a few lines later define SIZE_MAX to INT32_MAX.
      //   - SCO OpenServer (x86) headers typedef size_t to unsigned int
      //     but define SIZE_MAX to INT32_MAX.
      #if defined(__INTERIX) || defined(_SCO_DS)
      #	undef SIZE_MAX
      #endif

      // The code currently assumes that size_t is either 32-bit or 64-bit.
      #ifndef SIZE_MAX
      #	if SIZEOF_SIZE_T == 4
      #		define SIZE_MAX UINT32_MAX
      #	elif SIZEOF_SIZE_T == 8
      #		define SIZE_MAX UINT64_MAX
      #	else
      #		error size_t is not 32-bit or 64-bit
      #	endif
      #endif
      #if SIZE_MAX != UINT32_MAX && SIZE_MAX != UINT64_MAX
      #	error size_t is not 32-bit or 64-bit
      #endif
      ```

      There is a way to solve this to prevent compilation on non-32- or 64-bit platforms without requiring a `configure` script, but it involves a preprocessor macro to emit uncompilable code if a certain condition is met, which produces an error message unrelated to the actual failing condition (though, in practice, the macro name and its arguments are shown, so having something like `FAIL_BUILD_IF(sizeof(size_t) < 4)` makes it pretty obvious why the build failed). It basically involves inserting something like the following code somewhere, maybe in `cbits/lzma_wrapper.c` before the C code which provides the Haskell wrapper:

      ```c
      #define FAIL_BUILD_IF(condition) ((void)sizeof(char[1 - 2*!!(condition)]))

      // This function should never be called, it's simply a container so that
      // `FAIL_BUILD_IF` can emit uncompilable code into it.
      void _fail_build_if() {
        // The code currently assumes that size_t is either 32-bit or 64-bit.
        FAIL_BUILD_IF(sizeof(size_t) < 4);
      #		define SIZEOF_SIZE_T 4
        FAIL_BUILD_IF(sizeof(size_t) > 4 && sizeof(size_t) < 8);
      #		undef SIZEOF_SIZE_T
      #		define SIZEOF_SIZE_T 8
        FAIL_BUILD_IF(sizeof(size_t) > 8);

      // Incorrect(?) SIZE_MAX:
      //   - Interix headers typedef size_t to unsigned long,
      //     but a few lines later define SIZE_MAX to INT32_MAX.
      //   - SCO OpenServer (x86) headers typedef size_t to unsigned int
      //     but define SIZE_MAX to INT32_MAX.
      #if defined(__INTERIX) || defined(_SCO_DS)
      #	undef SIZE_MAX
      #endif

      #	ifndef SIZE_MAX
      #		if SIZEOF_SIZE_T == 8
      #			define SIZE_MAX UINT64_MAX
      #		elif SIZEOF_SIZE_T == 4
      #			define SIZE_MAX UINT32_MAX
      #		else
      #			error size_t is not 32-bit or 64-bit
      #		endif
      #	endif
      #	if SIZE_MAX != UINT32_MAX && SIZE_MAX != UINT64_MAX
      #		error size_t is not 32-bit or 64-bit
      #	endif
      }
      ```

      Apart from the `SIZEOF_SIZE_T` issue, a generic `config.h` which lacks `HAVE__BOOL` & `HAVE_STDBOOL_H` causes the following compilation error:

      ```c
      cbits/common/sysdefs.h:144:23: error:
           error: cannot combine with previous 'char' declaration specifier
          |
      144 | typedef unsigned char _Bool;
          |                       ^
      typedef unsigned char _Bool;
                            ^

      cbits/common/sysdefs.h:144:1: error:
           warning: typedef requires a name [-Wmissing-declarations]
          |
      144 | typedef unsigned char _Bool;
          | ^
      ```

      This has a simple fix:

      ```diff
      diff -ru a/cbits/common/sysdefs.h b/cbits/common/sysdefs.h
      --- a/cbits/common/sysdefs.h
      +++ b/cbits/common/sysdefs.h
      @@ -140,10 +140,10 @@
       #ifdef HAVE_STDBOOL_H
       #	include <stdbool.h>
       #else
      +#	define bool _Bool
       #	if ! HAVE__BOOL
      -typedef unsigned char _Bool;
      +typedef unsigned char bool;
       #	endif
      -#	define bool _Bool
       #	define false 0
       #	define true 1
       #	define __bool_true_false_are_defined 1
      ```
  - GHCup itself refuses to run on any platform it doesn't explicitly recognise.
    - Solution 1: add OpenBSD as a supported platform in `lib/GHCup/Types.hs`, and `lib/GHCup/Platform.hs`.
    - Solution 2: like with `lzma-static`'s `liblzma`, I feel it would be best to simply allow it to build and run on any platform (which GHC & GHCup themselves can run on). Is there a reason GHCup doesn't do that?
  - As mentioned, bindists. Building GHC on OpenBSD has some special considerations.
    - OpenBSD tar doesn't support the `-T` flag, used by `binary-dist` Make target.
    - GNU Make is required.
    - OpenBSD's `autoconf` & `automake` are shell scripts which delegate to the versions specified by `AUTOCONF_VERSION` & `AUTOMAKE_VERSION`.
    - `ln` doesn't recognise the `-v` flag, so I use this shim:
    - Some older GHC & OpenBSD versions also require patching. Some OpenBSD versions have different package or symlink requirements. I have some build scripts for some older versions:
      - GHC 8.2.2 on OpenBSD 6.6:

        ```diff
        diff -ru a/libraries/Cabal/Cabal/Distribution/PackageDescription/Configuration.hs b/libraries/Cabal/Cabal/Distribution/PackageDescription/Configuration.hs
        --- a/libraries/Cabal/Cabal/Distribution/PackageDescription/Configuration.hs
        +++ b/libraries/Cabal/Cabal/Distribution/PackageDescription/Configuration.hs
        @@ -56,7 +56,8 @@
         import Distribution.Types.Condition
         import Distribution.Types.DependencyMap
         
        -import qualified Data.Map as Map
        +import qualified Data.Map.Lazy as Map
        +import qualified Data.Map.Strict as Map.Strict
         import Data.Tree ( Tree(Node) )
         
         ------------------------------------------------------------------------------
        @@ -219,7 +220,7 @@
             mp m@(Right _) _           = m
             mp _           m@(Right _) = m
             mp (Left xs)   (Left ys)   =
        -        let union = Map.foldrWithKey (Map.insertWith' combine)
        +        let union = Map.foldrWithKey (Map.Strict.insertWith combine)
                             (unDepMapUnion xs) (unDepMapUnion ys)
                     combine x y = simplifyVersionRange $ unionVersionRanges x y
                 in union `seq` Left (DepMapUnion union)
        ```
      - GHC 8.6.5 on OpenBSD 6.6:

        ```diff
        diff -ru a/rts/RtsSymbols.c b/rts/RtsSymbols.c
        --- a/rts/RtsSymbols.c
        +++ b/rts/RtsSymbols.c
        @@ -278,9 +278,7 @@
         #endif
         
         #if defined(openbsd_HOST_OS)
        -#define RTS_OPENBSD_ONLY_SYMBOLS                            \
        -     SymE_NeedsProto(__guard_local)                         \
        -     SymE_NeedsProto(_DYNAMIC)
        +#define RTS_OPENBSD_ONLY_SYMBOLS SymE_NeedsProto(__guard_local)
         #else
         #define RTS_OPENBSD_ONLY_SYMBOLS
         #endif
        ```
      - GHC 8.8.4 on OpenBSD 6.8:
        - ```diff
          diff -ru a/utils/iserv/ghc.mk b/utils/iserv/ghc.mk
          --- a/utils/iserv/ghc.mk
          +++ b/utils/iserv/ghc.mk
          @@ -32,7 +32,7 @@
           # need it there's no alternative.
           # Don't do this on FreeBSD to work around #17962.
           ifeq "$(TargetElf)" "YES"
          -ifeq "$(findstring $(TargetOS_CPP), solaris2 freebsd)" ""
          +ifeq "$(findstring $(TargetOS_CPP), solaris2 freebsd openbsd)" ""
           # The Solaris linker does not support --export-dynamic option. It also
           # does not need it since it exports all dynamic symbols by default
           utils/iserv_stage2_MORE_HC_OPTS += -optl-Wl,--export-dynamic
          ```
        - GCC must be installed.
        - `python3` is not symlinked to `python3.$MINOR` by default.

--- a/cabal.project.release
+++ b/cabal.project.release
@@ -2,9 +2,10 @@ import: cabal.project.common
 
 optimization: 2
 
--- The release project file always wants to build with -tar.
--- The tar flag is only there to circumvent complicated errors during
--- development, which sometimes happens due to libarchive.
+-- The release project file always wants to build with -tar on the
+-- platforms it supports (which are those it provides config.h files
+-- for). The tar flag is only there to circumvent complicated errors
+-- during development, which sometimes happens due to libarchive.
 package ghcup
     flags: +tui
 
@@ -32,3 +33,6 @@ elif os(freebsd)
                xz -system-xz
   package *
     ghc-options: -split-sections
+else
+  package ghcup
+    flags: +tar

--- a/lib-opt/GHCup/OptParse/ChangeLog.hs
+++ b/lib-opt/GHCup/OptParse/ChangeLog.hs
@@ -136,6 +136,7 @@ changelog ChangeLogOptions{..} runAppState runLogger = do
               Darwin  -> exec "open" [T.unpack $ decUTF8Safe $ serializeURIRef' uri] Nothing Nothing
               Linux _ -> exec "xdg-open" [T.unpack $ decUTF8Safe $ serializeURIRef' uri] Nothing Nothing
               FreeBSD -> exec "xdg-open" [T.unpack $ decUTF8Safe $ serializeURIRef' uri] Nothing Nothing
+              OpenBSD -> exec "xdg-open" [T.unpack $ decUTF8Safe $ serializeURIRef' uri] Nothing Nothing
               Windows -> do
                 let args = "start \"\" " ++ (T.unpack $ decUTF8Safe $ serializeURIRef' uri)
                 c <- liftIO $ system $ args

--- a/lib/GHCup/Platform.hs
+++ b/lib/GHCup/Platform.hs
@@ -119,11 +119,17 @@ getPlatform = do
         either (const Nothing) Just . versioning . decUTF8Safe'
           <$> getFreeBSDVersion
       pure $ PlatformResult { _platform = FreeBSD, _distroVersion = ver }
+    "openbsd" -> do
+      ver <-
+        either (const Nothing) Just . versioning . decUTF8Safe'
+          <$> getOpenBSDVersion
+      pure $ PlatformResult { _platform = OpenBSD, _distroVersion = ver }
     "mingw32" -> pure PlatformResult { _platform = Windows, _distroVersion = Nothing }
     what -> throwE $ NoCompatiblePlatform what
   lift $ logDebug $ "Identified Platform as: " <> T.pack (prettyShow pfr)
   pure pfr
  where
+  getOpenBSDVersion = lift $ fmap _stdOut $ executeOut "uname" ["-r"] Nothing
   getFreeBSDVersion = lift $ fmap _stdOut $ executeOut "freebsd-version" [] Nothing
   getDarwinVersion = lift $ fmap _stdOut $ executeOut "sw_vers"
                                                         ["-productVersion"]
@@ -306,6 +312,7 @@ getStackGhcBuilds PlatformResult{..} = do
             [] -> []
             _ -> L.intercalate "-" c)
           libComponents
+      OpenBSD -> pure []
       FreeBSD ->
         case _distroVersion of
           Just fVer
@@ -343,6 +350,8 @@ getStackOSKey PlatformRequest { .. } =
     (A_64   , Darwin ) -> pure "macosx"
     (A_32   , FreeBSD) -> pure "freebsd32"
     (A_64   , FreeBSD) -> pure "freebsd64"
+    (A_32   , OpenBSD) -> pure "openbsd32"
+    (A_64   , OpenBSD) -> pure "openbsd64"
     (A_32   , Windows) -> pure "windows32"
     (A_64   , Windows) -> pure "windows64"
     (A_ARM  , Linux _) -> pure "linux-armv7"
@@ -350,6 +359,7 @@ getStackOSKey PlatformRequest { .. } =
     (A_Sparc, Linux _) -> pure "linux-sparc"
     (A_ARM64, Darwin ) -> pure "macosx-aarch64"
     (A_ARM64, FreeBSD) -> pure "freebsd-aarch64"
+    (A_ARM64, OpenBSD) -> pure "openbsd-aarch64"
     (arch', os') -> throwE $ UnsupportedSetupCombo arch' os'
 
 getStackPlatformKey :: (MonadReader env m, MonadFail m, HasLog env, MonadCatch m, MonadIO m)

--- a/lib/GHCup/Types.hs
+++ b/lib/GHCup/Types.hs
@@ -227,6 +227,7 @@ data Platform = Linux LinuxDistro
               | Darwin
               -- ^ must exit
               | FreeBSD
+              | OpenBSD
               | Windows
               -- ^ must exit
   deriving (Eq, GHC.Generic, Ord, Show)
@@ -237,6 +238,7 @@ platformToString :: Platform -> String
 platformToString (Linux distro) = "linux-" ++ distroToString distro
 platformToString Darwin = "darwin"
 platformToString FreeBSD = "freebsd"
+platformToString OpenBSD = "openbsd"
 platformToString Windows = "windows"
 
 instance Pretty Platform where

--- a/lib/GHCup/Types/JSON.hs
+++ b/lib/GHCup/Types/JSON.hs
@@ -144,12 +144,14 @@ instance ToJSONKey Platform where
     FreeBSD -> T.pack "FreeBSD"
     Linux (OtherLinux s) -> T.pack ("Linux_" <> s)
     Linux d -> T.pack ("Linux_" <> show d)
+    OpenBSD -> T.pack "OpenBSD"
     Windows -> T.pack "Windows"
 
 instance FromJSONKey Platform where
   fromJSONKey = FromJSONKeyTextParser $ \t -> if
     | T.pack "Darwin" == t -> pure Darwin
     | T.pack "FreeBSD" == t -> pure FreeBSD
+    | T.pack "OpenBSD" == t -> pure OpenBSD
     | T.pack "Windows" == t -> pure Windows
     | T.pack "Linux_" `T.isPrefixOf` t -> case
         T.stripPrefix (T.pack "Linux_") t

